linalg/batchlinalg.cpp

```diff
@@ -0,0 +1,391 @@
+
+   if (init)
+   {
+      init        = false;
+      auto status = MFEM_cu_or_hip(blasCreate)(&handle);
+      MFEM_VERIFY(status == MFEM_CU_or_HIP(BLAS_STATUS_SUCCESS),
+                  "Cannot initialize GPU BLAS");
+      atexit([]()
+      {
+         MFEM_cu_or_hip(blasDestroy)(handle);
+         handle = nullptr;
+      });
+   }
+   return handle;
+}
+#endif
+
+
+LibBatchSolver::LibBatchSolver(const DenseTensor &MatrixBatch,
+                               const SolveMode mode)
+   : mode_(mode)
+     // TODO: should this really be a copy?
+   , LUMatrixBatch_(MatrixBatch)
+{
+   if (!setup_)
+   {
+      Setup();
+   }
+}
+
+LibBatchSolver::LibBatchSolver(const SolveMode mode) : mode_(mode) {}
+
+void LibBatchSolver::AssignMatrices(const mfem::DenseTensor &MatrixBatch)
+{
+   // TODO: should this really be a copy?
+   LUMatrixBatch_ = MatrixBatch;
+   setup_         = false;  //Setup is now false as the matrices have changed
+   lu_valid_      = false;
+
+   //Need to always call setup since the matrices have changed
+   if (!setup_)
+   {
+      Setup();
+   }
+}
+
+void LibBatchSolver::AssignMatrices(const mfem::Vector &vMatrixBatch,
+                                    const int ndofs,
+                                    const int num_matrices)
+{
+   const int totalSize = ndofs * ndofs * num_matrices;
+   LUMatrixBatch_.SetSize(ndofs, ndofs,
+                          num_matrices); //needs to be temporary memory
+   double *d_LUMatrixBatch      = LUMatrixBatch_.Write();
+   const double *d_vMatrixBatch = vMatrixBatch.Read();
+
+   mfem::forall(totalSize, [=] MFEM_HOST_DEVICE (int i) { d_LUMatrixBatch[i] = d_vMatrixBatch[i]; });
+
+   AssignMatrices(LUMatrixBatch_);
+}
+
+void LibBatchSolver::GetInverse(mfem::DenseTensor &InvMatBatch) const
+{
+
+   /*
+    if (mode_ == SolveMode::INVERSE)
+    {
+       MFEM_WARNING("GetInverse with SolveMode::Inverse involves and extra memory copy, consider "
+                    "GetInverse(M, M_inv) instead");
+    }
+   */
+
+   if (!setup_)
+   {
+      mfem_error("LibBatchSolver has not been setup");
+   }
+
+   // use existing inverse
+   if (mode_ == SolveMode::INVERSE)
+   {
+      MFEM_VERIFY(InvMatrixBatch_.TotalSize() == InvMatBatch.TotalSize(),
+                  "Internal error, InvMatrixBatch_.TotalSize() != InvMatBatch.TotalSize()");
+
+      const double *d_M_inv = InvMatrixBatch_.Read();
+      double *d_out         = InvMatBatch.Write();
+
+      mfem::forall(InvMatrixBatch_.TotalSize(), [=] MFEM_HOST_DEVICE (int i) { d_out[i] = d_M_inv[i]; });
+   }
+   else if (mode_ == SolveMode::LU)
+   {
+      return ComputeInverse(InvMatBatch);
+   }
+   else
+   {
+      mfem_error("unsupported mode");
+   }
+}
+
+void LibBatchSolver::ComputeLU()
+{
+   if (lu_valid_)
+   {
+      return;
+   }
+
+#if defined(MFEM_USE_CUDA_OR_HIP)
+   if (Device::Allows(Backend::DEVICE_MASK))
+   {
+
+      mfem::Array<int> info_array(num_matrices_); // need to move to temp mem
```

Comment:
Can you add an optional `MemoryType` to the constructor and use that throughout? You can default to `MemoryType::DEFAULT`

Reply:
yes